File: public/methods/10_FORECAST_EVALUATION/QM005_MODEL_CONFIDENCE_SET/labs/python/qm005_hands_on.py
"""Reader-facing QM005 lab: run a transparent Tmax Model Confidence Set."""
from __future__ import annotations
from pathlib import Path
import argparse
import numpy as np
import pandas as pd

HERE=Path(__file__).resolve(); ROOT=HERE.parents[2]
DATA=ROOT/'figures/source_data/qm005_losses.csv'
MODELS=['A','B','C','D','E']

def load_indices(n:int)->np.ndarray:
    path=ROOT/f'figures/source_data/qm005_bootstrap_indices_n{n}.csv'
    if not path.exists(): raise SystemExit('This lab includes fixed bootstrap indices for --sample 80 or 240.')
    return pd.read_csv(path).to_numpy(dtype=int)-1

def mcs_tmax(L:pd.DataFrame,idx:np.ndarray,alpha:float):
    names=list(L.columns); Xall=L.to_numpy(float); active=list(range(len(names))); rows=[]
    while len(active)>1:
        X=Xall[:,active]; means=X.mean(0); rel=means-means.mean()
        bm=X[idx].mean(axis=1); br=bm-bm.mean(axis=1,keepdims=True); se=br.std(axis=0,ddof=1)
        t=rel/se; stat=float(t.max()); bs=((br-rel)/se).max(axis=1)
        p=float((1+np.sum(bs>=stat))/(len(bs)+1)); j=int(np.argmax(t)); elim=active[j]
        rows.append((len(active),names[elim],stat,p,p<=alpha))
        if p>alpha: break
        active.remove(elim)
    return [names[i] for i in active],rows

def main():
    ap=argparse.ArgumentParser(); ap.add_argument('--sample',type=int,choices=[80,240],default=240); ap.add_argument('--alpha',type=float,default=0.10)
    a=ap.parse_args(); d=pd.read_csv(DATA).iloc[:a.sample]; idx=load_indices(a.sample)
    final,rows=mcs_tmax(d[MODELS],idx,a.alpha)
    print(f'Sample: {a.sample} observations | confidence level: {100*(1-a.alpha):.1f}%')
    print('\nMean squared loss:')
    for m,v in d[MODELS].mean().items(): print(f'  {m}: {v:.3f}')
    print('\nSequential MCS path:')
    for k,(nm,elim,stat,p,reject) in enumerate(rows,1):
        action=f'eliminate {elim}' if reject else 'stop'
        print(f'  step {k}: {nm} models | Tmax={stat:.3f} | p={p:.3f} | {action}')
    print('Final MCS: {'+', '.join(final)+'}')
    print(f'QM005_LAB_FINAL_SET={",".join(final)}')
    for k,(nm,elim,stat,p,reject) in enumerate(rows,1): print(f'QM005_LAB_STEP{k}_P={p:.12f}')

if __name__=='__main__': main()

File: public/methods/10_FORECAST_EVALUATION/QM005_MODEL_CONFIDENCE_SET/labs/r/qm005_hands_on.R
# Reader-facing QM005 lab. Usage: Rscript qm005_hands_on.R 240 0.10
args <- commandArgs(trailingOnly=TRUE); n <- ifelse(length(args)>=1,as.integer(args[1]),240); alpha <- ifelse(length(args)>=2,as.numeric(args[2]),0.10)
if (!(n %in% c(80,240))) stop("This lab includes fixed bootstrap indices for sample sizes 80 or 240.")
full <- commandArgs(trailingOnly=FALSE); f <- grep("^--file=",full,value=TRUE); script <- normalizePath(sub("^--file=","",f[1]),winslash="/")
root <- normalizePath(file.path(dirname(script),"..",".."),winslash="/",mustWork=TRUE)
d <- read.csv(file.path(root,"figures","source_data","qm005_losses.csv"),check.names=FALSE)[seq_len(n),]
idx <- read.csv(file.path(root,"figures","source_data",sprintf("qm005_bootstrap_indices_n%d.csv",n)),check.names=FALSE)
models <- c("A","B","C","D","E")
active <- seq_along(models); step <- 1; rows <- list()
repeat {
  if (length(active)<=1) break
  X <- as.matrix(d[,models[active],drop=FALSE]); means <- colMeans(X); rel <- means-mean(means)
  bm <- matrix(NA_real_,nrow=nrow(idx),ncol=length(active))
  for (b in seq_len(nrow(idx))) bm[b,] <- colMeans(X[as.integer(idx[b,]),,drop=FALSE])
  br <- bm-rowMeans(bm); se <- apply(br,2,sd); tt <- rel/se; stat <- max(tt)
  centered <- sweep(sweep(br,2,rel,"-"),2,se,"/"); bs <- apply(centered,1,max); p <- (1+sum(bs>=stat))/(nrow(idx)+1)
  j <- which.max(tt); elim <- active[j]; rows[[step]] <- c(n_models=length(active),tmax=stat,p=p,eliminate=elim)
  if (p>alpha) break
  active <- active[-j]; step <- step+1
}
cat(sprintf("Sample: %d observations | confidence level: %.1f%%\n",n,100*(1-alpha)))
cat(sprintf("Final MCS: {%s}\n",paste(models[active],collapse=", ")))
cat(sprintf("QM005_R_LAB_FINAL_SET=%s\n",paste(models[active],collapse=",")))
